### ISSUE TITLE
Support of water heaters

### ISSUE THREAD

**Issue by Bascht74** [esphome/feature-requests#2047]
**Describe the problem you have/What new integration you would like**
Home Assistant has an integration for water heaters: 
https://www.home-assistant.io/integrations/water_heater/

I would like to use them within ESPHome as well. 
They are similar to climate devices but don't control a room temperature, but a water temperature

**Please describe your use case for this integration and alternatives you've tried:**
Control my water heater that is connected to Mobus via modbus controller as a template water heater

**Additional context**
./.

**Comment by randybb** [comment]
I think it is used only for the heaters that are directly communicating with HA.
I am controlling my heater/water heater via esphome and I am using the standard climate component (one component for heater and one for water heater).

**Comment by MiguelAngelLV** [comment]
Home Assistant will add new tiles card for Water Heater. I think could be very interesting have a especific water heater type in EspHome

**Comment by barndawgie** [comment]
For the esphome-econet integration, we currently use a Climate entity to represent a Water Heater. Being able to use the HA Water Heater entity would be interesting for sure!

**Comment by kpfleming** [comment]
I've begun working on this, PRs will be linked to this feature request.

Water heater support in HA is extremely basic at the moment, it's not possible to report the actual heating mode of a hybrid unit, for example.

**Comment by barndawgie** [comment]
> Water heater support in HA is extremely basic at the moment, it's not possible to report the actual heating mode of a hybrid unit, for example.

Yes, I've been looking at it and am honestly not sure we'd move ESPHome-EcoNet as I think we'd lose a bunch of functionality with the current HA implementation.

**Comment by kpfleming** [comment]
I can't tell for sure, but I don't think there would be much functionality loss. The water_heater platform in HA supports many of the necessary modes (electric, heat_pump, etc.) which are currently 'custom presets' in ESPHome-EcoNet. It's certainly worth having a conversation on the issue over there to document which bits of functionality are needed.

**Comment by dhoeben** [comment]
Hi all, I just created a pull request regarding this. Hopefully it will pass and feel free to check it out and let me know where to improve.

**Comment by lookasek** [comment]
> Hi all, I just created a pull request regarding this. Hopefully it will pass and feel free to check it out and let me know where to improve.

@dhoeben are you planning to prepare a new pull request with water heater component? I see that the current PR status is Closed.

**Comment by dhoeben** [comment]
> > Hi all, I just created a pull request regarding this. Hopefully it will pass and feel free to check it out and let me know where to improve.
> 
> @dhoeben are you planning to prepare a new pull request with water heater component? I see that the current PR status is Closed.

Yep, but very busy at the moment. I think I will publish beginning next year, I will keep you updated!

**Comment by dhoeben** [comment]
I just created a new PR
